python/dgl/subgraph.py

```diff
@@ -0,0 +1,519 @@
+    """
+    etype_ids = [graph.get_etype_id(etype) for etype in etypes]
+    # meta graph is homograph, still using int64
+    meta_src, meta_dst, _ = graph._graph.metagraph.find_edges(utils.toindex(etype_ids, "int64"))
+    rel_graphs = [graph._graph.get_relation_graph(i) for i in etype_ids]
+    meta_src = meta_src.tonumpy()
+    meta_dst = meta_dst.tonumpy()
+    ntypes_invmap = {n: i for i, n in enumerate(set(meta_src) | set(meta_dst))}
+    mapped_meta_src = [ntypes_invmap[v] for v in meta_src]
+    mapped_meta_dst = [ntypes_invmap[v] for v in meta_dst]
+    node_frames = [graph._node_frames[i] for i in ntypes_invmap]
+    edge_frames = [graph._edge_frames[i] for i in etype_ids]
+    induced_ntypes = [graph._ntypes[i] for i in ntypes_invmap]
+    induced_etypes = [graph._etypes[i] for i in etype_ids]   # get the "name" of edge type
+    num_nodes_per_induced_type = [graph.number_of_nodes(ntype) for ntype in induced_ntypes]
+
+    metagraph = graph_index.from_edge_list((mapped_meta_src, mapped_meta_dst), True)
+    # num_nodes_per_type should be int64
+    hgidx = heterograph_index.create_heterograph_from_relations(
+        metagraph, rel_graphs, utils.toindex(num_nodes_per_induced_type, "int64"))
+    hg = DGLHeteroGraph(hgidx, induced_ntypes, induced_etypes, node_frames, edge_frames)
+    return hg
+
+DGLHeteroGraph.edge_type_subgraph = edge_type_subgraph
+
+#################### Internal functions ####################
+
+def _create_hetero_subgraph(parent, sgi, induced_nodes, induced_edges):
+    """Internal function to create a subgraph.
+
+    Parameters
+    ----------
+    parent : DGLGraph
+        The parent DGLGraph.
+    sgi : HeteroSubgraphIndex
+        Subgraph object returned by CAPI.
+    induced_nodes : list[Tensor] or None
+        Induced node IDs. Will store it as the dgl.NID ndata unless it
+        is None, which means the induced node IDs are the same as the parent node IDs.
+    induced_edges : list[Tensor] or None
+        Induced edge IDs. Will store it as the dgl.EID ndata unless it
+        is None, which means the induced edge IDs are the same as the parent edge IDs.
+
+    Returns
+    -------
+    DGLGraph
+        Graph
+    """
+    if induced_nodes is None:
+        node_frames = [nf.clone() for nf in parent._node_frames]
+    else:
+        node_frames = []
+        for i, ind_nodes in enumerate(induced_nodes):
+            subf = parent._node_frames[i].subframe(ind_nodes)
+            subf[NID] = ind_nodes
+            node_frames.append(subf)
+    if induced_edges is None:
+        edge_frames = [nf.clone() for nf in parent._edge_frames]
+    else:
+        edge_frames = []
+        for i, ind_edges in enumerate(induced_edges):
+            subf = parent._edge_frames[i].subframe(ind_edges)
+            subf[EID] = ind_edges
+            edge_frames.append(subf)
+
+    hsg = DGLHeteroGraph(sgi.graph, parent.ntypes, parent.etypes,
+                         node_frames, edge_frames)
+    return hsg
```

Comment:
I'm wondering if it is possible to have a utility function that let a user automatically copy-on-read features from a parent graph given the induced NID and EID.  For instance, we would also want to auto-slice features in various transformation functions like `sample_neighbors`, `select_topk`, `compact_graphs`, and `to_block`.

Reply:
> I'm wondering if it is possible to have a (public) utility function that let a user automatically copy-on-read features from a parent graph given the induced NID and EID.

Can we instead implement them using subgraph routines such as `edge_subgraph`?

Reply:
I partially take my claim back: we probably don't need a public utility function, but I think it's still better to have an internal utility function, since `compact_graphs` and `to_block` are not generating subgraphs but they need the copy-on-read functionality.

`sample_neighbors` and `select_topk` generates subgraphs and they can probably reuse `_create_hetero_subgraph`.

What PinSage samples is not a subgraph, but I guess it's OK to just update it with `neighbor_graph.ndata.update(self.G.nodes[ntype].data)` since they have the same node space.

Reply:
Made a utility function called `utils.extract_subframes`.